File: spring-web-java/src/main/java/spring/web/java/presentation/exception/BaseException.java
package spring.web.java.presentation.exception;

import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ProblemDetail;
import org.springframework.web.ErrorResponse;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor
public class BaseException extends RuntimeException implements ErrorResponse {

	private final ErrorCode errorCode;
	private final HttpStatus httpStatus;

	@Override
	public HttpStatusCode getStatusCode() {
		return httpStatus;
	}

	@Override
	public ProblemDetail getBody() {
		ProblemDetail problemDetail = ProblemDetail.forStatus(httpStatus);
		problemDetail.setTitle(httpStatus.getReasonPhrase());
		problemDetail.setDetail(errorCode.getDetail());

		return problemDetail;
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/presentation/exception/BaseException.kt
package spring.web.kotlin.presentation.exception

import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import org.springframework.web.ErrorResponse

class BaseException(
    val errorCode: ErrorCode,
    val httpStatus: HttpStatus
) : RuntimeException(), ErrorResponse {
    override fun getStatusCode() = httpStatus

    override fun getBody() = ProblemDetail.forStatus(httpStatus).apply {
        title = httpStatus.reasonPhrase
        detail = errorCode.detail
    }
}